Fixed source model tracker's behavior on SET notifications.

NiklasRentzCAU opened the pull request:
Fixes #6.

sailingKieler commented:
Sorry, for my late answer.
I have one concern that is the mismatch with the alternative implementation of `IPropertyHolder` in https://github.com/eclipse/elk/blob/master/plugins/org.eclipse.elk.graph/src/org/eclipse/elk/graph/properties/MapPropertyHolder.java
Could that be a problem? `MapPropertyHolder` doesn't allow to listener for notifications so probably not.

The second concern is, that it won't be possible to explicitly set the default, will it?
(Could affect exported kgraph files provided by customers for debugging purposes)
Therefore, I suggest to add an additional check like `&& property.getDefaultValue() == null` in `EMapPropertyHolder`.

Last but not least I ask you to not hijack the existing `ViewContextSourceModelTrackingTest`, but add an additional test class, or at least duplicate the particular tests in the test class.
Since most of the customers rely on the `SimpleUpdateStrategy` testing of that combination should be continued.

sailingKieler commented on plugins/de.cau.cs.kieler.klighd.kgraph/src/de/cau/cs/kieler/klighd/kgraph/impl/EMapPropertyHolderImpl.java:
```diff
@@ -174,7 +174,7 @@ public void makePersistent() {
         if (value == null) {
             getProperties().removeKey(property);
-        } else {
+        } else if (getProperties().get(property) != value) {
```
IMO `|| property.getDefaultValue() != null` would be better here.

sailingKieler replied:
I just realized that you're referring to the properties map directly instead of going via `getProperty(...)`, so default values won't be an issue, right?

NiklasRentzCAU commented:
In the KGraph variant of the property handling, the notification is sent by the `IPropertyToObjectMapImpl`, ELK has a class that seems to do something similar and that also uses notifications [here](https://github.com/eclipse/elk/blob/4e545a5100eb647e94e8b52c34e37c327e262019/plugins/org.eclipse.elk.graph/src/org/eclipse/elk/graph/impl/ElkPropertyToValueMapEntryImpl.java#L125).
Do you think the notification is necessary if we set a value to the same value it was before? If so, I could also think of another solution for this problem digging into the code of the incremental update to not use the `IPropertyHolder`'s method `copyProperties` [here](https://github.com/kieler/KLighD/blob/a32ce9dcc2b2fbc902d61e3d4a2215abcab5353d/plugins/de.cau.cs.kieler.klighd.incremental/src/de/cau/cs/kieler/klighd/incremental/merge/KGraphMerger.java#L373), knowing that would break the source model tracker inside it. That however also feels like a hack, as the `copyProperties` method should just work and copy all properties, not remove some random adapter used in it.

And i would not call the new test I added hijacking the existing `ViewContextSourceModelTrackingTest` as I added another test for specifically that adapter that was breaking before. As you can see, all previous tests still use the simple update strategy, I just added the one test using the incremental update strategy and adapted your helping methods to allow for both, so that I do not need to duplicate code. The incremental update will get an own test class at some point, but this one test tests an issue that only happens with the combination of the source model tracker and the incremental update, so I thought I might as well reuse the code of that class that I would need anyway and put the test there. If you think I should put that test in an own test class, I can do that as well, but this addition changes nothing your costumers rely on.

NiklasRentzCAU commented:
@sailingKieler I changed the fix to the better working and more consistent solution as discussed last week. Please have a quick look at this so we can get this into master.
